util/ranger/ranger_test.go

```diff
@@ -1765,3 +1768,431 @@ create table t(
+	})
+}
+
+func TestShardIndexFuncSuites(t *testing.T) {
+
+	dom, store, err := newDomainStoreWithBootstrap(t)
+	defer func() {
+		dom.Close()
+		require.NoError(t, store.Close())
+	}()
+	require.NoError(t, err)
+	testKit := testkit.NewTestKit(t, store)
+	sctx := testKit.Session().(sessionctx.Context)
+
+	// -------------------------------------------
+	// test IsValidShardIndex function
+	// -------------------------------------------
+	longlongType := types.NewFieldType(mysql.TypeLonglong)
+	col0 := &expression.Column{UniqueID: 0, ID: 0, RetType: types.NewFieldType(mysql.TypeLonglong)}
+	col1 := &expression.Column{UniqueID: 1, ID: 1, RetType: types.NewFieldType(mysql.TypeLonglong)}
+	// col2 is GC column and VirtualExpr = tidb_shard(col0)
+	col2 := &expression.Column{UniqueID: 2, ID: 2, RetType: types.NewFieldType(mysql.TypeLonglong)}
+	col2.VirtualExpr = expression.NewFunctionInternal(sctx, ast.TidbShard, col2.RetType, col0)
+	// col3 is GC column and VirtualExpr = abs(col0)
+	col3 := &expression.Column{UniqueID: 3, ID: 3, RetType: types.NewFieldType(mysql.TypeLonglong)}
+	col3.VirtualExpr = expression.NewFunctionInternal(sctx, ast.Abs, col2.RetType, col0)
+	col4 := &expression.Column{UniqueID: 4, ID: 4, RetType: types.NewFieldType(mysql.TypeLonglong)}
+
+	cols := []*expression.Column{col0, col1}
+
+	// input is nil
+	require.False(t, ranger.IsValidShardIndex(nil))
+	// only 1 column
+	require.False(t, ranger.IsValidShardIndex([]*expression.Column{col2}))
+	// first col is not expression
+	require.False(t, ranger.IsValidShardIndex(cols))
+	// field in tidb_shard is not the secondary column
+	require.False(t, ranger.IsValidShardIndex([]*expression.Column{col2, col1}))
+	// expressioin is abs that is not tidb_shard
+	require.False(t, ranger.IsValidShardIndex([]*expression.Column{col3, col0}))
+	// normal case
+	require.True(t, ranger.IsValidShardIndex([]*expression.Column{col2, col0}))
+
+	// -------------------------------------------
+	// test IsEqOrInFunc function
+	// -------------------------------------------
+	// input is nil
+	require.False(t, ranger.IsEqOrInFunc(nil))
+	// input is column
+	require.False(t, ranger.IsEqOrInFunc(col0))
+	// input is abs
+	require.False(t, ranger.IsEqOrInFunc(col3.VirtualExpr))
+
+	// normal case
+	con1 := &expression.Constant{Value: types.NewDatum(1), RetType: longlongType}
+	con5 := &expression.Constant{Value: types.NewDatum(5), RetType: longlongType}
+	exprEq := expression.NewFunctionInternal(sctx, ast.EQ, col0.RetType, col0, con1)
+	exprIn := expression.NewFunctionInternal(sctx, ast.In, col0.RetType, col0, con1, con5)
+	require.NotNil(t, exprEq)
+	require.NotNil(t, exprIn)
+	require.True(t, ranger.IsEqOrInFunc(exprEq))
+	require.True(t, ranger.IsEqOrInFunc(exprIn))
+
+	// -------------------------------------------
+	// test ExtractColumnsFromExpr function
+	// -------------------------------------------
+	// input is nil
+	require.True(t, len(ranger.ExtractColumnsFromExpr(nil)) == 0)
```

Comment:
```suggestion
	require.Equal(t, len(ranger.ExtractColumnsFromExpr(nil)), 0)
```
Use `require.Equal` instead, it offers more information when test failed.